File: 4_median_of_two_sorted_arrays/solution.ts
function findMedianSortedArrays(nums1: number[], nums2: number[]): number {
  if (nums2.length < nums1.length) {
    ;[nums1, nums2] = [nums2, nums1]
  }

  let [left, right] = [0, nums1.length - 1]
  const totalLength = nums1.length + nums2.length
  const half = Math.floor(totalLength / 2)

  while (true) {
    const mid1 = Math.floor(left + (right - left) / 2)
    const mid2 = half - mid1 - 2

    // Handle the edge cases
    const getLeft = (arr: number[], index: number) =>
      index >= 0 ? arr[index] : -Infinity
    const getRight = (arr: number[], index: number) =>
      index + 1 < arr.length ? arr[index + 1] : Infinity

    const [leftA, leftB] = [getLeft(nums1, mid1), getLeft(nums2, mid2)]
    const [rightA, rightB] = [getRight(nums1, mid1), getRight(nums2, mid2)]

    if (leftA <= rightB && leftB <= rightA) {
      return totalLength % 2 === 0
        ? (Math.max(leftA, leftB) + Math.min(rightA, rightB)) / 2
        : Math.min(rightA, rightB)
    } else if (rightB < leftA) {
      right = mid1 - 1
      // when rightA < leftB
    } else {
      left = mid1 + 1
    }
  }
}

File: 4_median_of_two_sorted_arrays/solution.py
from typing import List


class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        A, B = nums1, nums2
        total = len(A) + len(B)
        half = total // 2

        if len(A) > len(B):
            A, B = B, A

        l, r = 0, len(A) - 1

        while True:
            i = l + (r - l) // 2
            j = half - i - 2

            ALeft = A[i] if i >= 0 else float("-inf")
            ARight = A[i + 1] if (i + 1) < len(A) else float("inf")
            BLeft = B[j] if j >= 0 else float("-inf")
            BRight = B[j + 1] if (j + 1) < len(B) else float("inf")

            # found the correct partition
            if ALeft <= BRight and BLeft <= ARight:
                if total % 2 == 1:
                    return min(ARight, BRight)
                else:
                    leftMax = max(ALeft, BLeft)
                    rightMin = min(ARight, BRight)
                    return (leftMax + rightMin) / 2
            elif ALeft > BRight:
                r = i - 1
            else:
                l = i + 1
